netCDF4/_netCDF4.pyx

```diff
@@ -3754,27 +3757,64 @@ rename a `netCDF4.Variable` attribute named `oldname` to `newname`."""
             # note: missing_value has to have same endian-ness as variable
             # or this won't work.
             mval = numpy.array(self.missing_value, self.dtype)
+            # create mask from missing values. 
+            mvalmask = numpy.zeros(data.shape, numpy.bool)
+            # set mask=True for data outside valid_min,valid_max.
+            # (issue #576)
+            validmin = None; validmax = None
+            # if valid_range exists use that, otherwise
+            # look for valid_min, valid_max.  No special
+            # treatment of byte data as described at
+            # http://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html).
+            if hasattr(self, 'valid_range') and len(self.valid_range) == 2:
+                validmin = numpy.array(self.valid_range[0], self.dtype)
+                validmax = numpy.array(self.valid_range[1], self.dtype)
+            else:
+                if hasattr(self, 'valid_min'):
+                    validmin = numpy.array(self.valid_min, self.dtype)
+                if hasattr(self, 'valid_max'):
+                    validmax = numpy.array(self.valid_max, self.dtype)
+            # http://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html).
+            # "If the data type is byte and _FillValue 
+            # is not explicitly defined,
+            # then the valid range should include all possible values.
+            # Otherwise, the valid range should exclude the _FillValue
+            # (whether defined explicitly or by default) as follows. 
+            # If the _FillValue is positive then it defines a valid maximum,
+            #  otherwise it defines a valid minimum."
+            byte_type = self.dtype.str[1:] in ['u1','i1']
+            if hasattr(self, '_FillValue'):
+                fval = numpy.array(self._FillValue, self.dtype)
+            else:
+                fval = numpy.array(default_fillvals[self.dtype.str[1:]],self.dtype)
+                if byte_type: fval = 0.
+            if validmin is None and fval < 0:
```

Comment:
I like this logic, and it does conform to the spec, but I wonder what happens if a _FillValue is defined as 0? It's obvously a bit obscure, but not inconcevable. My reading of it is that a _FillValue of 0 should be a validmin. Perhaps default fval to None?